framework/src/modules/interoperability/utils.ts

```diff
@@ -106,3 +146,295 @@ export const computeValidatorsHash = (
+		return {
+			status: VerifyStatus.FAIL,
+			error: new Error(
+				`Sending partner chain ${txParams.sendingChainID} is in registered status so certificate cannot be empty.`,
+			),
+		};
+	}
+
+	return {
+		status: VerifyStatus.OK,
+	};
+};
+
+export const checkCertificateValidity = (
+	partnerChainAccount: ChainAccount,
+	encodedCertificate: Buffer,
+): VerificationResult => {
+	if (!encodedCertificate.equals(EMPTY_BYTES)) {
+		const decodedCertificate = codec.decode<Certificate>(certificateSchema, encodedCertificate);
+		if (
+			decodedCertificate.blockID.equals(EMPTY_BYTES) ||
+			decodedCertificate.stateRoot.equals(EMPTY_BYTES) ||
+			decodedCertificate.validatorsHash.equals(EMPTY_BYTES) ||
+			decodedCertificate.timestamp === 0
+		) {
+			return {
+				status: VerifyStatus.FAIL,
+				error: new Error('Certificate is missing required values.'),
+			};
+		}
+
+		// Last certificate height should be less than new certificate height
+		if (decodedCertificate.height <= partnerChainAccount.lastCertificate.height) {
+			return {
+				status: VerifyStatus.FAIL,
+				error: new Error('Certificate height should be greater than last certificate height.'),
+			};
+		}
+	}
+
+	return {
+		status: VerifyStatus.OK,
+	};
+};
+
+export const checkActiveValidatorsUpdate = (
+	txParams: CrossChainUpdateTransactionParams,
+): VerificationResult => {
+	if (
+		txParams.activeValidatorsUpdate.length !== 0 ||
+		txParams.newCertificateThreshold > BigInt(0)
+	) {
+		// Non-empty certificate
+		if (txParams.certificate.equals(EMPTY_BYTES)) {
+			return {
+				status: VerifyStatus.FAIL,
+				error: new Error(
+					'Certificate cannot be empty when activeValidatorsUpdate is non-empty or newCertificateThreshold >0.',
+				),
+			};
+		}
+
+		// params.activeValidatorsUpdate has the correct format
+		for (let i = 0; i < txParams.activeValidatorsUpdate.length; i += 1) {
+			const currentValidator = txParams.activeValidatorsUpdate[i];
+			const nextValidator = txParams.activeValidatorsUpdate[i + 1];
+			if (currentValidator.blsKey.byteLength !== VALID_BLS_KEY_LENGTH) {
+				return {
+					status: VerifyStatus.FAIL,
+					error: new Error(`BlsKey length should be equal to ${VALID_BLS_KEY_LENGTH}.`),
+				};
+			}
+			if (currentValidator.blsKey.compare(nextValidator.blsKey) > -1) {
+				return {
+					status: VerifyStatus.FAIL,
+					error: new Error('Validators blsKeys must be unique and lexicographically ordered'),
+				};
+			}
+		}
+	}
+
+	return {
+		status: VerifyStatus.OK,
+	};
+};
+
+export const checkInboxUpdateValidity = (
+	txParams: CrossChainUpdateTransactionParams,
+	partnerChannelData: ChannelData,
+): VerificationResult => {
+	const partnerChainIDBuffer = getIDAsKeyForStore(txParams.sendingChainID);
+	const decodedCertificate = codec.decode<Certificate>(certificateSchema, txParams.certificate);
+	const { crossChainMessages, messageWitness, outboxRootWitness } = txParams.inboxUpdate;
+	const ccmHashes = crossChainMessages.map(ccm => hash(ccm));
```

Comment:
Just to be sure, this should be hashed here?

Reply:
hmm not 100% sure but below is my understanding,
 
according to LIP0045 `appendPath: An array of hashes necessary to append new data to the tree efficiently. The default value of this property is an empty array.`

But I think when we receive it in InboxUpdate, we receive only serialized CCMs that are decoded and executed so we need to hash them to use it as appendPath to calculate the root

Reply:
Just adding an reference to confirm this, the data is always hashed before being appended to the tree : https://github.com/LiskHQ/lips/blob/main/proposals/lip-0045.md#execution-1